### PR TITLE
Propagate temporary lifetime extension into if and match.

### PR DESCRIPTION
**Pull request by m-ou-se**
This PR makes this work:

```rust
let a = if true {
    ..;
    &temp() // used to error, but now gets lifetime extended
} else {
    ..;
    &temp() // used to error, but now gets lifetime extended
};
```

and

```rust
let a = match () {
    _ => {
        ..;
        &temp() // used to error, but now gets lifetime extended
    }
};
```

to make it consistent with:

```rust
let a = {
    ..;
    &temp() // lifetime is extended
};
```

This is one small part of [the temporary lifetimes work](https://github.com/rust-lang/lang-team/issues/253).

This part is backwards compatible (so doesn't need be edition-gated), because all code affected by this change previously resulted in a hard error.

### CODE REVIEW

**Comment by m-ou-se** [comment]
cc @nikomatsakis @dingxiangfei2009 see https://rust-lang.zulipchat.com/#narrow/stream/403629-t-lang.2Ftemporary-lifetimes-2024/topic/Separate.20the.20part.20on.20if.20bodies.20and.20match.20arms.3F/near/422433608

File: tests/ui/lifetimes/temporary-lifetime-extension.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        &temp()
+    };
+    let g = match true {
+        true => &temp(),
+        false => {
+            let _ = 123;
+            &temp()
+        }
+    };
+    println!("{a:?} {b:?} {c:?} {d:?} {e:?} {f:?} {g:?}");
```
Comment (danielhenrymantilla) [inline tests/ui/lifetimes/temporary-lifetime-extension.rs:28]: I think the following (with, _e.g._, `::std::env::vars().nth(1)`), would be useful to ensure as well:

```suggestion
    };
    let h = match temp() {
        owned_non_temporary => &{ owned_non_temporary },
    };
    println!("{a:?} {b:?} {c:?} {d:?} {e:?} {f:?} {g:?} {h:?}");
```

**Review by danielhenrymantilla — APPROVED** [review]
Yesss!! Such an underrated improvement 🤩

**Comment by joshtriplett** [comment]
Based on discussion in the @rust-lang/lang meeting today:

@rfcbot merge

**Comment by kennytm** [comment]
:thinking:  do these make sense?

```rust
let x = loop { break &temp(); };
let x = 'a: { break 'a &temp(); };
```

**Comment by m-ou-se** [comment]
@kennytm Maybe, but that'd be a separate proposal. I'm keeping the scope of this PR minimal. :)

**Comment by m-ou-se** [comment]
r? compiler

**Comment by m-ou-se** [comment]
r? compiler

**Review by BoxyUwU — APPROVED** [review]

**Comment by BoxyUwU** [comment]
cool
r? @compiler-errors idk

**Review by compiler-errors — APPROVED** [review]

**Comment by CAD97** [comment]
This looks to be the likely cause of [a regression in temporary lifetime extension in constant initializers](https://github.com/rust-lang/rust/issues/126562).

**Comment by CAD97** [comment]
Although looking at the actual change, it doesn't quite make sense how this would be responsible when the other expr kinds don't "cover" the temporary lifetime the way that's being seen in #126562. So I guess there must be a different location where const promotion is handled that also needs to be taught to peek through `if` and `match` expression kinds.

**Comment by CAD97** [comment]
https://github.com/rust-lang/rust/blob/55cac26a9ef17da1c9c77c0816e88e178b7cc5dd/compiler/rustc_hir_typeck/src/rvalue_scopes.rs#L26-L52

**Comment by CAD97** [comment]
(see issue for further information; it looks like this may have gotten fingered incorrectly)

**Pull request merged**